File: app/src/main/java/mx/edu/upqroo/kristenandroid/data/repositories/KardexRepository.java
package mx.edu.upqroo.kristenandroid.data.repositories;

import android.app.Application;
import android.os.AsyncTask;

import java.util.List;

import androidx.lifecycle.LiveData;
import mx.edu.upqroo.kristenandroid.data.database.KristenRoomDatabase;
import mx.edu.upqroo.kristenandroid.data.database.daos.KardexDao;
import mx.edu.upqroo.kristenandroid.data.database.entities.Kardex;
import mx.edu.upqroo.kristenandroid.managers.SessionManager;
import mx.edu.upqroo.kristenandroid.api.sie.SieApiServices;

public class KardexRepository {
    private static KardexRepository mInstance;
    private KardexDao mKardexDao;
    private SieApiServices mApi;

    private KardexRepository(Application application) {
        KristenRoomDatabase db = KristenRoomDatabase.getInstance(application);
        mKardexDao = db.kardexDao();
        mApi = SieApiServices.getInstance(application);
    }

    public static KardexRepository getInstance(Application application) {
        if (mInstance == null) {
            mInstance = new KardexRepository(application);
        }
        return mInstance;
    }

    public LiveData<List<Kardex>> getKardex(String userId) {
        LiveData<List<Kardex>> response = mKardexDao.getByUserId(userId);
        AsyncTask.execute(() -> {
            if (mKardexDao.count() == 0) {
                mApi.getKardexList(SessionManager.getInstance().getSession().getUserId(),
                        SessionManager.getInstance().getSession().getConfig().getUserToken());
            }
        });
        return response;
    }

    public void updateKardexFromService() {
        mApi.getKardexList(SessionManager.getInstance().getSession().getUserId(),
                SessionManager.getInstance().getSession().getConfig().getUserToken());
    }

    public void insert(Kardex kardex) {
        AsyncTask.execute(() -> mKardexDao.insert(kardex));
    }

    public void deleteAll() {
        AsyncTask.execute(() -> mKardexDao.deleteAll());
    }

    public void update(Kardex... kardex){
        AsyncTask.execute(() -> mKardexDao.update(kardex));
    }
}

File: app/src/main/java/mx/edu/upqroo/kristenandroid/data/repositories/KardexRepository.kt
package mx.edu.upqroo.kristenandroid.data.repositories

import android.app.Application
import android.os.AsyncTask

import androidx.lifecycle.LiveData
import mx.edu.upqroo.kristenandroid.data.database.KristenRoomDatabase
import mx.edu.upqroo.kristenandroid.data.database.daos.KardexDao
import mx.edu.upqroo.kristenandroid.data.database.entities.Kardex
import mx.edu.upqroo.kristenandroid.managers.SessionManager
import mx.edu.upqroo.kristenandroid.api.sie.SieApiServices

class KardexRepository private constructor(application: Application) {
    private val mKardexDao: KardexDao
    private val mApi: SieApiServices

    init {
        val db = KristenRoomDatabase.getInstance(application)
        mKardexDao = db.kardexDao()
        mApi = SieApiServices.getInstance(application)
    }

    fun getKardex(userId: String): LiveData<List<Kardex>> {
        val response = mKardexDao.getByUserId(userId)
        AsyncTask.execute {
            if (mKardexDao.count() == 0) {
                mApi.getKardexList(SessionManager.instance.session.userId,
                        SessionManager.instance.session.config.userToken)
            }
        }
        return response
    }

    fun updateKardexFromService() {
        mApi.getKardexList(SessionManager.instance.session.userId,
                SessionManager.instance.session.config.userToken)
    }

    fun insert(kardex: Kardex) {
        AsyncTask.execute { mKardexDao.insert(kardex) }
    }

    fun deleteAll() {
        AsyncTask.execute { mKardexDao.deleteAll() }
    }

    fun update(vararg kardex: Kardex) {
        AsyncTask.execute { mKardexDao.update(*kardex) }
    }

    companion object {
        private var mInstance: KardexRepository? = null

        fun getInstance(application: Application): KardexRepository {
            if (mInstance == null) {
                mInstance = KardexRepository(application)
            }
            return mInstance as KardexRepository
        }
    }
}
